README.md
yum-dell Cookbook
======================

[![Build Status](https://secure.travis-ci.org/atomic-penguin/cookbook-yum-dell.png?branch=master)](http://travis-ci.org/atomic-penguin/cookbook-yum-dell)

Configures Dell community and hardware/OMSA (Open Manage Server Assistant)
repositories.  See http://linux.dell.com for full details.

The default recipe configures community and hardware repositories, and installs
OMSA on Dell hardware.  On non-Dell hardware, only the community repository is
configured.

The community repository provides non-Dell specific
dkms (Dynamic Kernel Module Support), and SMBIOS (Systems Management BIOS)
library packages.

The hardware repositories contain hardware drivers, firmware binaries,
BIOS updates, and OMSA software specific to Dell hardware.

Requirements
------------

There is a dependency on the dmidecode package so that Ohai can detect Dell
hardware.

This cookbook depends on the `yum_repository` provider from the `yum` cookbook.
You need to have a RHEL family platform, and yum, to use the cookbook.

#### cookbooks 

- `yum` - Opscode maintained 3.0+ cookbook 

Attributes
----------
The following are overridable attributes, in the `yum['dell']` namespace.

#### yum-dell::default

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['yum']['dell']['enabled']</tt></td>
    <td>Boolean</td>
    <td>Whether Dell hardware was detected via Ohai, or not.</td>
    <td><tt>Value depends on hardware.</tt></td>
  </tr>
  <tr>
    <td><tt>['yum']['dell']['packages']</tt></td>
    <td>Array</td>
    <td>An array of packages from Dell repositories to install.</td>
    <td><tt>srvadmin-all</tt> on Dell hardware. <tt>Empty</tt> on non-Dell hardware.</td>
  </tr>
</table>

Each Dell repository has its own attribute namespace.  It is recommended
that you not override these, unless you have a local mirror.  The attributes
correspond to the parameters in the `yum_repository` library.

#### per-repository attributes

<table>
  <tr>
    <th>Namespace</th>
    <th>Corresponding repository</th>
  </tr>
  <tr>
    <td><tt>yum['dell']['community']</tt></td>
    <td>Community supported Open Source software.</td>
  </tr>
  <tr>
    <td><tt>yum['dell']['omsa-indep']</tt></td>
    <td>Dell OMSA repository, Hardware independent</td>
  </tr>
  <tr>
    <td><tt>yum['dell']['omsa-specific']</tt></td>
    <td>Dell OMSA repository, Hardware specific</td>
  </tr>
</table>

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>$namespace['repositoryid']</tt></td>
    <td>String</td>
    <td>Filename of repo file.</td>
    <td><tt>dell-community</tt>, <tt>dell-omsa-indep</tt>, <tt>dell-omsa-specific</tt> respectively</td>
  </tr>
  <tr>
    <td><tt>$namespace['description']</tt></td>
    <td>String</td>
    <td>Human readable description for repository.</td>
    <td>e.g. Dell Community Repository</td>
  </tr>
  <tr>
    <td><tt>$namespace['mirrorlist']</tt></td>
    <td>String</td>
    <td>URL with list of mirrors.</td>
    <td>repository specific</td>
  </tr>
  <tr>
    <td><tt>$namespace['gpgkey']</tt></td>
    <td>String</td>
    <td>URL of public signing key for package</td>
    <td>repository specific</td>
  </tr>
  <tr>
    <td><tt>$namespace['gpgcheck']</tt></td>
    <td>Boolean</td>
    <td>Whether, or not, to check the provided gpgkey</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>$namespace['failovermethod']</tt></td>
    <td>List</td>
    <td>Method to choose next mirror on failure.  Either, priority or roundrobin.</td>
    <td><tt>priority</tt></td>
  </tr>
</table>

Usage
-----
#### yum-corporate::default

Optionally, set attributes in a role, and
include `yum-dell` in your node's `run_list`:

```
default_attributes(
  :yum => {
    :dell => {
      :packages => [ "srvadmin-base" ]
    }
  }
)
```

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[yum-dell]"
  ]
}
```

#### yum-dell::undo

This recipe does the opposite action of yum-corporate::default.

Specifically, this recipe runs srvadmin-uninstall.sh to remove
all OMSA comoponents.  It deletes all Dell repository `.repo`
files.  Finally, it purges any leftover settings left after
uninstall in `/opt/dell/srvadmin`.

Contributing
------------

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------

Author:: Eric G. Wolfe
Copyright:: 2010-2011

Author:: Tippr, Inc.
Copyright:: 2010

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
yum-dell Cookbook CHANGELOG
======================
This file is used to list changes made in each version of the yum-dell cookbook.

v3.0.5
------

* Update GPG key URL
* Constrain yum dependency to 3.x or higher 

v3.0.4
------

* Remove yum_key workaround
* Update tests
* Remove stale key file
* Change kitchen plugin to docker
* Add Jenkins task for internal testing

v3.0.3
------

* Fix i686 URI
* Add regression tests for kernel.machine i686 case.

v3.0.2
------

* Update tests
* Workaround for COOK-4164

v3.0.1
------

Update metadata

v3.0.0
------

Reworked against [yum](http://github.com/opscode-cookbooks/yum) 3.0.0

* Split dell recipe out from soon to be deprecated, yumrepo
  - Changed variable scope from `yumrepo['dell']` to `yum['dell']`
  - Updated kitchen files
  - Added ChefSpec unit tests
  - Removed minitest unit tests
  - Added bats integration tests
  - Added Rubocop style guide linter
  - Added undo recipe

v2.1.0
------

* Pin against yum < 3.0.0

v2.0.0
------

* Add test-kitchen skeleton
  - Validated recipes against yum v2.3.0
* Fix minitest-spec tests for Chef 11
* Deprecated annvix repo
  - Appears to be no longer maintained for EL6

v1.0.0
------

* vmware-tools
  - Bump version to ESX 5.1
  - re-instated to default recipe
  - add upstart support for vmware-tools 9.0 packages on EL 6
  - Change LWRP action to create (updates repo file), instead of add
* deprecate/remove epel recipe
  - has been moved over to Opscodes yum::epel
* disable minitest spec for Travis
  - This isn't working on TravisCI, will revisit unit tests later

v0.16.6
-------

* add zeromq repo
  - contributed by @BryanWB
* Bump chef gem in Gemfile
  - Locked chef gem at 10.X, 'cause minitest spec seems to be broke
    on chef 11.x.
  
TODO - Replace tests with chefspec

v0.16.5
-------

Fix mirrorlist attribute in Dell recipe.

v0.16.4
-------

Thanks to Paul Graydon (@twirrim) for adding Percona MySQL repositories.

Thanks to Bryan Berry (@bryanwb) for adding yet another PostgreSQL
repository. :)

Breaking changes (postgresql)

* Recipe renamed, postgresql9 -> postgresql.
* Added node['repo']['postgresql']['version'] to select version of
  Postgres via attribute, from pgrpms.org repository.

v0.16.3
-------

Set name/base_url to localdomain in the case Ohai `domain` is
unavailable.  Fixes chef-solo compilation error on Vagrant boxes.

v0.16.2
-------

    Change corporate recipe to key off domain name.

v0.16.0
-------

    Add VMWare Tools for ESX v5.0u1, along with a vmware-tools-upgrade
recipe.  The yumrepo::vmware-tools recipe is backwards incompatible.
Unfortunately there is no way to upgrade vmware-tools rpm files without
yanking them out, and replacing them.

    One must ensure the `vmware-tools-upgrade` recipe will run **before**
the `vmware-tools` recipe.  Otherwise the 4.1 recipe will directly
conflict with the 5.0u1 version of the recipe.

    I also added some attribute sanity tests for this new version, enjoy!


spec/undo_spec.rb
require 'spec_helper'

describe 'yum-dell::undo' do
  context 'on Centos 6.4' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'centos', version: 6.4, step_into: ['yum_repository']).converge(described_recipe)
    end

    it 'deletes yum repository dell-community' do
      expect(chef_run).to delete_yum_repository('dell-community')
    end

    it 'deletes yum repository dell-omsa-indep' do
      expect(chef_run).to delete_yum_repository('dell-omsa-indep')
    end

    it 'deletes yum repository dell-omsa-specific' do
      expect(chef_run).to delete_yum_repository('dell-omsa-specific')
    end

    it 'recursively deletes /opt/dell/srvadmin' do
      expect(chef_run).to delete_directory('/opt/dell/srvadmin').with(
        recursive: true
      )
    end
  end
end


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

RSpec.configure do |config|
  config.formatter = :documentation
end

at_exit { ChefSpec::Coverage.report! }


spec/default_spec.rb
require 'spec_helper'

describe 'yum-dell::default' do
  context 'on Centos 6.4 x86_64 with Dell hardware' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'centos', version: 6.4, step_into: ['yum_repository']) do |node|
        node.automatic['dmi']['system']['manufacturer'] = 'Dell Inc.'
        node.automatic['kernel']['machine'] = 'x86_64'
      end.converge(described_recipe)
    end

    it 'installs dmidecode' do
      expect(chef_run).to install_package('dmidecode')
    end

    it 'creates dell-community repository with correct mirrorlist' do
      expect(chef_run).to create_yum_repository('dell-community').with(
        mirrorlist: 'http://linux.dell.com/repo/community/mirrors.cgi?osname=el6&basearch=$basearch'
      )
    end

    it 'creates dell-omsa-indep repository with correct mirrorlist' do
      expect(chef_run).to create_yum_repository('dell-omsa-indep').with(
        mirrorlist: 'http://linux.dell.com/repo/hardware/latest/mirrors.cgi?osname=el6&basearch=$basearch&native=1&dellsysidpluginver=$dellsysidpluginver'
      )
    end

    it 'renders dell-omsa-specific repository with correct mirrorlist' do
      expect(chef_run).to create_yum_repository('dell-omsa-specific').with(
        mirrorlist: 'http://linux.dell.com/repo/hardware/latest/mirrors.cgi?osname=el6&basearch=$basearch&native=1&sys_ven_id=$sys_ven_id&sys_dev_id=$sys_dev_id&dellsysidpluginver=$dellsysidpluginver'
      )
    end

    it 'installs srvadmin-all' do
      expect(chef_run).to install_package('srvadmin-all')
    end
  end

  context 'on Centos 6.4 x86_64 on non-Dell hardware' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'centos', version: 6.4, step_into: ['yum_repository']) do |node|
        node.automatic['dmi']['system']['manufacturer'] = 'VMWare Inc.'
        node.automatic['kernel']['machine'] = 'x86_64'
      end.converge(described_recipe)
    end

    it 'creates dell-community repository with correct mirrorlist' do
      expect(chef_run).to create_yum_repository('dell-community').with(
        mirrorlist: 'http://linux.dell.com/repo/community/mirrors.cgi?osname=el6&basearch=$basearch'
      )
    end

    it 'to not create dell-omsa-indep repository' do
      expect(chef_run).to_not create_yum_repository('dell-omsa-indep')
    end

    it 'to not create dell-omsa-specific repository' do
      expect(chef_run).to_not create_yum_repository('dell-omsa-specific')
    end

    it 'to not install srvadmin-all' do
      expect(chef_run).to_not install_package('srvadmin-all')
    end
  end

  context 'on Centos 6.4 i686 on Dell hardware' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'centos', version: 6.4, step_into: ['yum_repository']) do |node|
        node.automatic['dmi']['system']['manufacturer'] = 'Dell Inc.'
        node.automatic['kernel']['machine'] = 'i686'
      end.converge(described_recipe)
    end

    it 'creates dell-community repository with i386 architecture' do
      expect(chef_run).to create_yum_repository('dell-community').with(
        mirrorlist: 'http://linux.dell.com/repo/community/mirrors.cgi?osname=el6&basearch=i386',
        gpgkey: 'http://linux.dell.com/repo/community/content/el6-i386/repodata/repomd.xml.key'
      )
    end

    it 'creates dell-omsa-specific repository with i386 architecture' do
      expect(chef_run).to create_yum_repository('dell-omsa-specific').with(
        mirrorlist: 'http://linux.dell.com/repo/hardware/latest/mirrors.cgi?osname=el6&basearch=i386&native=1&sys_ven_id=$sys_ven_id&sys_dev_id=$sys_dev_id&dellsysidpluginver=$dellsysidpluginver'
      )
    end
  end
end


recipes/undo.rb
#
# Cookbook Name:: yum-dell
# Recipe:: undo
#
# Copyright 2010, Eric G. Wolfe
# Copyright 2010, Tippr Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Force uninstall of all srvadmin packages.
execute 'srvadmin-uninstall-force' do
  command '/opt/dell/srvadmin/sbin/srvadmin-uninstall.sh -f'
  action :run
  only_if { ::File.exists?('/opt/dell/srvadmin/sbin/srvadmin-uninstall.sh') }
end

# Delete Dell repository files.
%w[ community omsa-indep omsa-specific ].each do |repo|
  yum_repository node['yum']['dell'][repo]['repositoryid'] do
    action :delete
  end
end

# Unlink /opt/dell/srvadmin
directory '/opt/dell/srvadmin' do
  recursive true
  action :delete
end

unless Chef::Config[:solo]
  ruby_block 'remove yum-dell::undo from run_list when there is a conflict' do
    block do
      node.run_list.remove('recipe[yum-dell::undo]')
    end
    only_if { node.run_list.include?('recipe[yum-dell::default]') }
  end
end


recipes/default.rb
#
# Cookbook Name:: yum-dell
# Recipe:: default
#
# Copyright 2010, Eric G. Wolfe
# Copyright 2010, Tippr Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Detection of Dell hardware depends on dmidecode
package 'dmidecode'

# Community supported software.  Does not require Dell Hardware.
yum_repository node['yum']['dell']['community']['repositoryid'] do
  description node['yum']['dell']['community']['description']
  mirrorlist node['yum']['dell']['community']['mirrorlist']
  gpgkey node['yum']['dell']['community']['gpgkey']
  gpgcheck node['yum']['dell']['community']['gpgcheck']
  failovermethod node['yum']['dell']['community']['failovermethod']
  action :create
end

# Dell Hardware/OMSA repositories.  Requires Dell Hardware.
%w[omsa-indep omsa-specific].each do |repo|
  yum_repository node['yum']['dell'][repo]['repositoryid'] do
    description node['yum']['dell'][repo]['description']
    mirrorlist node['yum']['dell'][repo]['mirrorlist']
    gpgkey node['yum']['dell'][repo]['gpgkey']
    gpgcheck node['yum']['dell'][repo]['gpgcheck']
    failovermethod node['yum']['dell'][repo]['failovermethod']
    only_if { node['yum']['dell']['enabled'] }
  end
end

# Install packages, srvadmin-all by default
node['yum']['dell']['packages'].each do |dell_package|
  package dell_package
end


attributes/omsa-indep.rb
#
# Cookbook Name:: yum-dell
# Attributes:: omsa-indep
#
# Copyright 2010, Eric G. Wolfe
# Copyright 2010, Tippr Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default['yum']['dell']['omsa-indep']['repositoryid'] = 'dell-omsa-indep'
default['yum']['dell']['omsa-indep']['description'] = 'Dell OMSA repository - Hardware independent'
case node['kernel']['machine']
when /i[3456]86/
  default['yum']['dell']['omsa-indep']['mirrorlist'] = 'http://linux.dell.com/repo/hardware/latest/mirrors.cgi?' +
    "osname=el#{node['platform_version'].to_i}&basearch=i386&native=1&dellsysidpluginver=$dellsysidpluginver"
else
  default['yum']['dell']['omsa-indep']['mirrorlist'] = 'http://linux.dell.com/repo/hardware/latest/mirrors.cgi?' +
    "osname=el#{node['platform_version'].to_i}&basearch=$basearch&native=1&dellsysidpluginver=$dellsysidpluginver"
end
default['yum']['dell']['omsa-indep']['gpgkey'] = 'http://linux.dell.com/repo/hardware/latest/public.key'
default['yum']['dell']['omsa-indep']['gpgcheck'] = true
default['yum']['dell']['omsa-indep']['failovermethod'] = 'priority'


attributes/omsa-specific.rb
#
# Cookbook Name:: yum-dell
# Attributes:: omsa-specific
#
# Copyright 2010, Eric G. Wolfe
# Copyright 2010, Tippr Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default['yum']['dell']['omsa-specific']['repositoryid'] = 'dell-omsa-specific'
default['yum']['dell']['omsa-specific']['description'] = 'Dell OMSA repository - Hardware specific'
case node['kernel']['machine']
when /i[3456]86/
  default['yum']['dell']['omsa-specific']['mirrorlist'] = 'http://linux.dell.com/repo/hardware/latest/mirrors.cgi?' +
    "osname=el#{node['platform_version'].to_i}&basearch=i386&native=1&sys_ven_id=$sys_ven_id&sys_dev_id=$sys_dev_id&dellsysidpluginver=$dellsysidpluginver"
else
  default['yum']['dell']['omsa-specific']['mirrorlist'] = 'http://linux.dell.com/repo/hardware/latest/mirrors.cgi?' +
    "osname=el#{node['platform_version'].to_i}&basearch=$basearch&native=1&sys_ven_id=$sys_ven_id&sys_dev_id=$sys_dev_id&dellsysidpluginver=$dellsysidpluginver"
end
default['yum']['dell']['omsa-specific']['gpgkey'] = 'http://linux.dell.com/repo/hardware/latest/public.key'
default['yum']['dell']['omsa-specific']['gpgcheck'] = true
default['yum']['dell']['omsa-specific']['failovermethod'] = 'priority'
default['yum']['dell']['omsa-specific']['packages'] = %w[ srvadmin-all ]


attributes/community.rb
#
# Cookbook Name:: yum-dell
# Attributes:: community
#
# Copyright 2010, Eric G. Wolfe
# Copyright 2010, Tippr Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default['yum']['dell']['community']['repositoryid'] = 'dell-community'
default['yum']['dell']['community']['description'] = 'Dell Community Repository'
case node['kernel']['machine']
when /i[3456]86/
  default['yum']['dell']['community']['mirrorlist'] = 'http://linux.dell.com/repo/community/mirrors.cgi?' +
    "osname=el#{node['platform_version'].to_i}&basearch=i386"
  default['yum']['dell']['community']['gpgkey'] = 'http://linux.dell.com/repo/community/content/' +
    "el#{node['platform_version'].to_i}-i386/repodata/repomd.xml.key"
else
  default['yum']['dell']['community']['mirrorlist'] = 'http://linux.dell.com/repo/community/mirrors.cgi?' +
    "osname=el#{node['platform_version'].to_i}&basearch=$basearch"
  default['yum']['dell']['community']['gpgkey'] = 'http://linux.dell.com/repo/community/content/' +
    "el#{node['platform_version'].to_i}-#{node['kernel']['machine']}/repodata/repomd.xml.key"
end
default['yum']['dell']['community']['gpgcheck'] = true
default['yum']['dell']['community']['failovermethod'] = 'priority'
